### PR TITLE
IIR refactor I

### PR DESCRIPTION
**Pull request by cosunae**
Technical Description
==================

* restructure all the IIR data structures into an IIR subproject and provide a first version of the IIRNode (however it changed significantly infurther PRs of these series)
* Add an `iir` namespace

### CODE REVIEW

File: src/dawn/CodeGen/GridTools/GTCodeGen.cpp [hunk 1]
```diff
@@ -882,13 +886,13 @@ std::string GTCodeGen::generateStencilInstantiation(
...
-        RangeToString(", ", "(dom, ",
-                      ")")(stencils[i]->getFields(false), [&](const Stencil::FieldInfo& field) {
-          if(stencilInstantiation->isAllocatedField(field.AccessID))
-            return "m_" + field.Name;
-          else
-            return field.Name;
-        }));
+        RangeToString(", ", "(dom, ", ")")(
+            stencils[i]->getFields(false), [&](const iir::Stencil::FieldInfo& field) {
+              if(stencilInstantiation->isAllocatedField(field.AccessID))
```
Comment (twicki) [inline src/dawn/CodeGen/GridTools/GTCodeGen.cpp:891]: should we keep brackets around these types of stmts even though they are 1-liner?

File: src/dawn/IIR/DoMethod.cpp [hunk 2]
```diff
@@ -12,27 +12,28 @@
...
 namespace dawn {
+namespace iir {
 
 DoMethod::DoMethod(Interval interval)
     : interval_(interval), id_(IndexGenerator::Instance().getIndex()), dependencyGraph_(nullptr) {}
 
-std::vector<std::shared_ptr<StatementAccessesPair>>& DoMethod::getStatementAccessesPairs() {
-  return statementAccessesPairs_;
-}
+// std::vector<std::shared_ptr<StatementAccessesPair>>& DoMethod::getStatementAccessesPairs() {
```
Comment (twicki) [inline src/dawn/IIR/DoMethod.cpp:29]: remove comments

File: src/dawn/IIR/DoMethod.cpp [hunk 3]
```diff
@@ -103,7 +104,7 @@ class CheckNonNullStatementVisitor : public ASTVisitorForwarding, public NonCopy
 };
 
 bool DoMethod::isEmptyOrNullStmt() const {
-  for(auto const& statementAccessPair : statementAccessesPairs_) {
+  for(auto const& statementAccessPair : children_) {
```
Comment (twicki) [inline src/dawn/IIR/DoMethod.cpp:107]: 2/2: what is the pattern we want here?

I think we should always call getChildren

File: src/dawn/IIR/DoMethod.cpp [hunk 4]
```diff
@@ -49,7 +50,7 @@ std::shared_ptr<DependencyGraphAccesses>& DoMethod::getDependencyGraph() {
 boost::optional<Extents> DoMethod::computeMaximumExtents(const int accessID) const {
   boost::optional<Extents> extents;
 
-  for(auto& stmtAccess : getStatementAccessesPairs()) {
+  for(auto& stmtAccess : getChildren()) {
```
Comment (twicki) [inline src/dawn/IIR/DoMethod.cpp:53]: 1/2

File: src/dawn/IIR/DoMethod.h [hunk 5]
```diff
@@ -30,14 +32,16 @@ class StatementAccessesPair;
...
 ///
 /// @ingroup optimizer
-class DoMethod {
+class DoMethod : public IIRNode<Stage, DoMethod, StatementAccessesPair> {
   Interval interval_;
   const long unsigned int id_;
 
   std::shared_ptr<DependencyGraphAccesses> dependencyGraph_;
-  std::vector<std::shared_ptr<StatementAccessesPair>> statementAccessesPairs_;
+  //  std::vector<std::shared_ptr<StatementAccessesPair>> statementAccessesPairs_;
```
Comment (twicki) [inline src/dawn/IIR/DoMethod.h:40]: rm

File: src/dawn/IIR/DoMethod.h [hunk 6]
```diff
@@ -49,9 +53,9 @@ class DoMethod {
   DoMethod& operator=(DoMethod&&) = default;
   /// @}
 
-  /// @brief Get the statements of the Stage
-  std::vector<std::shared_ptr<StatementAccessesPair>>& getStatementAccessesPairs();
-  const std::vector<std::shared_ptr<StatementAccessesPair>>& getStatementAccessesPairs() const;
+  //  /// @brief Get the statements of the Stage
```
Comment (twicki) [inline src/dawn/IIR/DoMethod.h:56]: rm

File: src/dawn/IIR/Field.cpp [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+//
+//  This file is distributed under the MIT License (MIT).
+//  See LICENSE.txt for details.
+//
+//===------------------------------------------------------------------------------------------===//
+
+#include "dawn/IIR/Field.h"
+
+namespace dawn {
+// namespace iir {
```
Comment (twicki) [inline src/dawn/IIR/Field.cpp:18]: rm

File: src/dawn/IIR/Field.cpp [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+void mergeFields(std::unordered_map<int, Field> const& sFields,
```
Comment (twicki) [inline src/dawn/IIR/Field.cpp:20]: maybe write out source and destination?

File: src/dawn/IIR/IIRNode.h [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+namespace iir {
+
+namespace impl {
+template <typename T>
+using StdVector = std::vector<T, std::allocator<T>>;
+}
+
+template <typename P>
+bool ptrEqual(const std::weak_ptr<P>& f, const std::weak_ptr<P>& s) {
+  return !f.expired() && f.lock() == s.lock();
```
Comment (twicki) [inline src/dawn/IIR/IIRNode.h:38]: should the comparison not be `*(f.lock()) == *(s.lock())`? like this you're comparing shared ptrs

File: src/dawn/IIR/IIRNode.h [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+  template <typename TChild, typename TParent>
+  void insertChildImpl(TChild&& child,
+                       typename std::enable_if<!std::is_void<TParent>::value>::type* = 0) {
+    static_assert((std::is_same<TParent, Parent>::value), "not the same parent types");
+    setChildParent<TChild, Parent>(child);
+    children_.push_back(std::move(child));
+  }
+
+  template <typename TChild>
+  void insertChild(TChild&& child) {
```
Comment (twicki) [inline src/dawn/IIR/IIRNode.h:158]: why is this templated? can't we just use the Child type?

File: src/dawn/IIR/MultiStage.cpp [hunk 11]
```diff
@@ -35,7 +36,7 @@ MultiStage::split(std::deque<MultiStage::SplitIndex>& splitterIndices,
   std::vector<std::shared_ptr<MultiStage>> newMultiStages;
 
   int curStageIndex = 0;
-  auto curStageIt = stages_.begin();
+  auto curStageIt = children_.begin();
```
Comment (twicki) [inline src/dawn/IIR/MultiStage.cpp:39]: getChildren?

File: src/dawn/IIR/MultiStage.h [hunk 12]
```diff
@@ -87,6 +96,7 @@ class MultiStage {
   /// the new stages. This function consumes the input argument `splitterIndices`.
   ///
   /// @return New multi-stages
+  // TODO this should not be here
```
Comment (twicki) [inline src/dawn/IIR/MultiStage.h:99]: rm

File: src/dawn/IIR/Stage.cpp [hunk 13]
```diff
@@ -229,49 +228,41 @@ void Stage::update() {
...
+    for(const auto& statementAccessesPair : doMethod.getChildren()) {
       const auto& access = statementAccessesPair->getAccesses();
 
       // first => AccessID, second => Extent
       for(auto& accessPair : access->getWriteAccesses()) {
         if(!stencilInstantiation_.isField(accessPair.first))
           continue;
 
-        AccessIDToFieldMap[accessPair.first]->mergeWriteExtents(accessPair.second);
+        fields_.at(accessPair.first).mergeWriteExtents(accessPair.second);
```
Comment (twicki) [inline src/dawn/IIR/Stage.cpp:253]: why move to at from brackets? Because we deleted ctors?

File: src/dawn/IIR/Stage.cpp [hunk 14]
```diff
@@ -229,49 +228,41 @@ void Stage::update() {
...
         if(!stencilInstantiation_.isField(accessPair.first))
           continue;
 
-        AccessIDToFieldMap[accessPair.first]->mergeReadExtents(accessPair.second);
+        fields_.at(accessPair.first).mergeReadExtents(accessPair.second);
       }
     }
   }
+  // TODO
+  //  getParent()->update();
```
Comment (twicki) [inline src/dawn/IIR/Stage.cpp:265]: assume this comes later?

File: src/dawn/IIR/Stage.h [hunk 15]
```diff
@@ -123,8 +125,8 @@ class Stage {
   /// `Input`
   ///
   /// The fields are computed during `Stage::update`.
-  std::vector<Field>& getFields() { return fields_; }
-  const std::vector<Field>& getFields() const { return fields_; }
+  //  std::unordered_map<int, Field>& getFields() { return fields_; }
```
Comment (twicki) [inline src/dawn/IIR/Stage.h:128]: rm

File: src/dawn/IIR/Stencil.cpp [hunk 16]
```diff
@@ -100,20 +116,21 @@ Stencil::Stencil(StencilInstantiation& stencilInstantiation,
...
   return intervals;
 }
 
 std::vector<Stencil::FieldInfo> Stencil::getFields(bool withTemporaries) const {
   std::set<int> fieldAccessIDs;
-  for(const auto& multistage : multistages_) {
-    for(const auto& stage : multistage->getStages()) {
-      for(const auto& field : stage->getFields()) {
-        fieldAccessIDs.insert(field.getAccessID());
+  for(const auto& multistage : children_) {
```
Comment (twicki) [inline src/dawn/IIR/Stencil.cpp:129]: one iterator?

File: src/dawn/IIR/Stencil.cpp [hunk 17]
```diff
@@ -100,20 +116,21 @@ Stencil::Stencil(StencilInstantiation& stencilInstantiation,
 
 std::unordered_set<Interval> Stencil::getIntervals() const {
   std::unordered_set<Interval> intervals;
-  for(const auto& multistage : multistages_)
-    for(const auto& stage : multistage->getStages())
-      for(const auto& doMethod : stage->getDoMethods())
+  for(const auto& multistage : children_)
```
Comment (twicki) [inline src/dawn/IIR/Stencil.cpp:119]: inconsistency, `childern_` vs `getChildren()`

File: src/dawn/IIR/Stencil.cpp [hunk 18]
```diff
@@ -139,8 +156,8 @@ std::vector<Stencil::FieldInfo> Stencil::getFields(bool withTemporaries) const {
 
 std::vector<std::string> Stencil::getGlobalVariables() const {
   std::set<int> globalVariableAccessIDs;
-  for(const auto& multistage : multistages_) {
-    for(const auto& stage : multistage->getStages()) {
+  for(const auto& multistage : children_) {
+    for(const auto& stage : multistage->getChildren()) {
```
Comment (twicki) [inline src/dawn/IIR/Stencil.cpp:160]: 1 iterator

File: src/dawn/Optimizer/PassDataLocalityMetric.cpp [hunk 19]
```diff
@@ -291,13 +292,13 @@ std::unordered_map<int, ReadWriteAccumulator> computeReadWriteAccessesMetricPerA
...
-computeReadWriteAccessesMetric(const std::shared_ptr<StencilInstantiation>& instantiation,
-                               const MultiStage& multiStage) {
+computeReadWriteAccessesMetric(const std::shared_ptr<iir::StencilInstantiation>& instantiation,
+                               const iir::MultiStage& multiStage) {
   ReadWriteCounter readWriteCounter(instantiation, multiStage);
 
-  for(const auto& stage : multiStage.getStages())
-    for(const auto& DoMethod : stage->getDoMethods())
-      for(const auto& statementAccessesPair : DoMethod->getStatementAccessesPairs()) {
+  for(const auto& stage : multiStage.getChildren())
```
Comment (twicki) [inline src/dawn/Optimizer/PassDataLocalityMetric.cpp:299]: 1 loop

File: src/dawn/Optimizer/PassFieldVersioning.h [hunk 20]
```diff
@@ -72,10 +74,11 @@ class PassFieldVersioning : public Pass {
   /// @param stencil    Current stencil
   /// @param stencil    Current Do-Method
   /// @param loopOrder  Current loop order of the stage
-  /// @param stageIdx   @b Linear index of the stage in the stencil
+  /// @param stageIdx   @b Lineaiir::DependencyGraphAccesses the stencil
```
Comment (twicki) [inline src/dawn/Optimizer/PassFieldVersioning.h:77]: whops?

File: src/dawn/Optimizer/PassInlining.cpp [hunk 21]
```diff
@@ -462,21 +463,21 @@ tryInlineStencilFunction(PassInlining::InlineStrategyKind strategy,
...
 
   DetectInlineCandiates inliner(strategy_, stencilInstantiation);
 
   // Iterate all statements (top -> bottom)
-  for(auto& stencilPtr : stencilInstantiation->getStencils()) {
-    for(auto& multiStagePtr : stencilPtr->getMultiStages()) {
-      for(auto& stagePtr : multiStagePtr->getStages()) {
-        Stage& stage = *stagePtr;
-        DoMethod& doMethod = stage.getSingleDoMethod();
+  for(const auto& stencilPtr : stencilInstantiation->getStencils()) {
```
Comment (twicki) [inline src/dawn/Optimizer/PassInlining.cpp:474]: 1 loop

File: src/dawn/Optimizer/PassStageMerger.cpp [hunk 22]
```diff
@@ -82,54 +80,56 @@ bool PassStageMerger::run(const std::shared_ptr<StencilInstantiation>& stencilIn
...
           // stage to merge
-          for(auto candiateStageIt = std::next(curStageIt);
-              candiateStageIt != multiStage.getStages().rend(); ++candiateStageIt) {
-            Stage& candiateStage = **candiateStageIt;
-            std::vector<std::unique_ptr<DoMethod>>& candiateDoMethods =
-                candiateStage.getDoMethods();
+          for(auto candidateStageIt = std::next(curStageIt);
+              candidateStageIt != multiStage.childrenREnd(); ++candidateStageIt) {
+            iir::Stage& candidateStage = **candidateStageIt;
+            //            std::vector<std::unique_ptr<iir::DoMethod>>& candiateDoMethods =
```
Comment (twicki) [inline src/dawn/Optimizer/PassStageMerger.cpp:94]: rm

File: src/dawn/Optimizer/PassTemporaryFirstAccess.cpp [hunk 23]
```diff
@@ -81,14 +81,13 @@ bool PassTemporaryFirstAccess::run(
     // {AccesID : (isFirstAccessWrite, Stmt)}
     std::unordered_map<int, std::pair<bool, std::shared_ptr<Stmt>>> accessMap;
 
-    for(auto& multiStagePtr : stencilPtr->getMultiStages()) {
-      for(auto& stagePtr : multiStagePtr->getStages()) {
-        DoMethod& doMethod = stagePtr->getSingleDoMethod();
+    for(const auto& multiStagePtr : stencilPtr->getChildren()) {
```
Comment (twicki) [inline src/dawn/Optimizer/PassTemporaryFirstAccess.cpp:84]: 1 loop

**Review by twicki — APPROVED** [review]

File: src/dawn/IIR/IIRNode.cpp [hunk 24]
```diff
@@ -0,0 +1,19 @@
+//===--------------------------------------------------------------------------------*- C++ -*-===//
```
Comment (twicki) [inline src/dawn/IIR/IIRNode.cpp:1]: why do we need this file?

File: src/dawn/IIR/DoMethod.cpp (hunk 3 above)
Comment (cosunae) [reply src/dawn/IIR/DoMethod.cpp:107]: yes, this was fixed and homogeneous in later PR

File: src/dawn/IIR/DoMethod.cpp (hunk 4 above)
Comment (cosunae) [reply src/dawn/IIR/DoMethod.cpp:53]: done

File: src/dawn/IIR/DoMethod.h (hunk 5 above)
Comment (cosunae) [reply src/dawn/IIR/DoMethod.h:40]: done

File: src/dawn/IIR/DoMethod.h (hunk 6 above)
Comment (cosunae) [reply src/dawn/IIR/DoMethod.h:56]: done

File: src/dawn/IIR/Field.cpp (hunk 7 above)
Comment (cosunae) [reply src/dawn/IIR/Field.cpp:18]: done

File: src/dawn/IIR/Field.cpp (hunk 8 above)
Comment (cosunae) [reply src/dawn/IIR/Field.cpp:20]: done

File: src/dawn/IIR/IIRNode.h (hunk 9 above)
Comment (cosunae) [reply src/dawn/IIR/IIRNode.h:38]: this was gone in later PR, since we moved to unique_ptr

File: src/dawn/IIR/IIRNode.h (hunk 10 above)
Comment (cosunae) [reply src/dawn/IIR/IIRNode.h:158]: indeed, it was fixed in later PR

File: src/dawn/IIR/MultiStage.cpp (hunk 11 above)
Comment (cosunae) [reply src/dawn/IIR/MultiStage.cpp:39]: children_ is a member of this class, we can just access it directly, isnt?

File: src/dawn/IIR/MultiStage.h (hunk 12 above)
Comment (cosunae) [reply src/dawn/IIR/MultiStage.h:99]: done

File: src/dawn/IIR/Stage.cpp (hunk 13 above)
Comment (cosunae) [reply src/dawn/IIR/Stage.cpp:253]: yes, exactly. We discussed in the past, we could have our own access At() that has a bound check as an assert, but not in runtime... but if we want it we should apply this pattern everywhere, out of this PR

File: src/dawn/IIR/Stage.cpp (hunk 14 above)
Comment (cosunae) [reply src/dawn/IIR/Stage.cpp:265]: yes

File: src/dawn/IIR/Stage.h (hunk 15 above)
Comment (cosunae) [reply src/dawn/IIR/Stage.h:128]: done

File: src/dawn/IIR/Stencil.cpp (hunk 17 above)
Comment (cosunae) [reply src/dawn/IIR/Stencil.cpp:119]: can you explain why would you use getChildren() over children_ ? 
The usual pattern was using getChildren() to have access to children_ from outside the class, but from the class we can just access the member.

File: src/dawn/IIR/Stencil.cpp (hunk 16 above)
Comment (cosunae) [reply src/dawn/IIR/Stencil.cpp:129]: what do you mean?

File: src/dawn/IIR/Stencil.cpp (hunk 18 above)
Comment (cosunae) [reply src/dawn/IIR/Stencil.cpp:160]: in later PR we are using iterateIIROver

File: src/dawn/Optimizer/PassFieldVersioning.h (hunk 20 above)
Comment (cosunae) [reply src/dawn/Optimizer/PassFieldVersioning.h:77]: fixed

File: src/dawn/Optimizer/PassInlining.cpp (hunk 21 above)
Comment (cosunae) [reply src/dawn/Optimizer/PassInlining.cpp:474]: done later

File: src/dawn/Optimizer/PassStageMerger.cpp (hunk 22 above)
Comment (cosunae) [reply src/dawn/Optimizer/PassStageMerger.cpp:94]: fixed

File: src/dawn/Optimizer/PassTemporaryFirstAccess.cpp (hunk 23 above)
Comment (cosunae) [reply src/dawn/Optimizer/PassTemporaryFirstAccess.cpp:84]: done later

File: src/dawn/IIR/IIRNode.cpp (hunk 24 above)
Comment (cosunae) [reply src/dawn/IIR/IIRNode.cpp:1]: it was removed later

File: src/dawn/IIR/MultiStage.cpp (hunk 11 above)
Comment (twicki) [reply src/dawn/IIR/MultiStage.cpp:39]: we're just not consistent in this PR. There is a mix between both usages. I would grep for both and make sure we're only using children_ from the class directly. I think this is the better approach than having the non-const getter create updates

File: src/dawn/IIR/Stage.cpp (hunk 13 above)
Comment (twicki) [reply src/dawn/IIR/Stage.cpp:253]: yeah, makes sense

File: src/dawn/IIR/Stencil.cpp (hunk 16 above)
Comment (twicki) [reply src/dawn/IIR/Stencil.cpp:129]: is removed with a later pr, just have an iterator over multiple levels of the tree

File: src/dawn/IIR/Stencil.cpp (hunk 17 above)
Comment (twicki) [reply src/dawn/IIR/Stencil.cpp:119]: see comment above

File: src/dawn/IIR/Stencil.cpp (hunk 18 above)
Comment (twicki) [reply src/dawn/IIR/Stencil.cpp:160]: yeah, I just wanted to tag occurrences so we see them when merging backward

**Pull request closed without merging**